lib/ansible/modules/cloud/amazon/iam_saml_federation.py

```diff
@@ -0,0 +1,242 @@
+    def _list_saml_providers(self):
+        return self.conn.list_saml_providers()
+
+    @AWSRetry.backoff(tries=3, delay=5)
+    def _get_saml_provider(self, arn):
+        return self.conn.get_saml_provider(SAMLProviderArn=arn)
+
+    @AWSRetry.backoff(tries=3, delay=5)
+    def _update_saml_provider(self, arn, metadata):
+        return self.conn.update_saml_provider(SAMLProviderArn=arn, SAMLMetadataDocument=metadata)
+
+    @AWSRetry.backoff(tries=3, delay=5)
+    def _create_saml_provider(self, metadata, name):
+        return self.conn.create_saml_provider(SAMLMetadataDocument=metadata, Name=name)
+
+    @AWSRetry.backoff(tries=3, delay=5)
+    def _delete_saml_provider(self, arn):
+        return self.conn.delete_saml_provider(SAMLProviderArn=arn)
+
+    def _get_provider_arn(self, name):
+        providers = self._list_saml_providers()
+        for p in providers['SAMLProviderList']:
+            provider_name = p['Arn'].split('/', 1)[1]
+            if name == provider_name:
+                return p['Arn']
+
+        return None
+
+    def create_or_update_saml_provider(self, name, metadata):
+        if not metadata:
+            self.module.fail_json(msg="saml_metadata_document must be defined for present state")
+
+        res = {'changed': False}
+        try:
+            arn = self._get_provider_arn(name)
+        except (botocore.exceptions.ValidationError, botocore.exceptions.ClientError) as e:
+            res['msg'] = str(e)
+            self.module.fail_json(**res)
+
+        if arn:  # see if metadata needs updating
+            try:
+                resp = self._get_saml_provider(arn)
+            except (botocore.exceptions.ValidationError, botocore.exceptions.ClientError) as e:
+                res['msg'] = str(e)
+                res['debug'] = [arn]
+                self.module.fail_json(**res)
+
+            if metadata.strip() != resp['SAMLMetadataDocument'].strip():
+                # provider needs updating
+                res['changed'] = True
+                if not self.module.check_mode:
+                    try:
+                        resp = self._update_saml_provider(arn, metadata)
+                        res['saml_provider_arn'] = resp['SAMLProviderArn']
+                    except botocore.exceptions.ClientError as e:
+                        res['msg'] = str(e)
+                        res['debug'] = [arn, metadata]
+                        self.module.fail_json(**res)
+
+        else:  # create
+            res['changed'] = True
+            if not self.module.check_mode:
+                try:
+                    resp = self._create_saml_provider(metadata, name)
+                    res['saml_provider_arn'] = resp['SAMLProviderArn']
+                except botocore.exceptions.ClientError as e:
+                    res['msg'] = str(e)
+                    res['debug'] = [name, metadata]
+                    self.module.fail_json(**res)
+
+        self.module.exit_json(**res)
```

Comment:
It would be nice to return the object you created too:

`res['provider] = self._get_saml_provider(self._get_provider_arn(name))`